File: services/commands/weeklySimulation/Neuron.py
from datetime import date
from math import exp

from forge.utils import get_utc_date


class Neuron:
    value: float = 0
    __R_in: float = 1
    __C: float = 10
    __last_updated_at: date = get_utc_date()

    def __init__(self, R_in, C):
        self.__R_in = R_in
        self.__C = C

    def decay(self, amount: float):
        self._update(0, amount, 1)
        self.__last_updated_at = get_utc_date()

    def charge(self, amount: float):
        self._update(1, amount, self.__R_in)

    def discharge(self, amount: float):
        self._update(-1, amount, self.__R_in)

    def _update(self, target: int, amount: float, resistance: float):
        self.value += (target - self.value) * \
            (1 - exp(-amount / (resistance * self.__C)))

File: services/rule_engine/src/main/java/models/Neuron.java
package models;

import com.mindsmiths.sdk.utils.Utils;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Data
@NoArgsConstructor
public class Neuron {
    private String id = Utils.randomString();

    private double value = 0;
    private double R_in = 1; // resistance
    private double C = 10; // capacity

    private LocalDateTime lastUpdatedAt = Utils.now();

    public Neuron(double R_in, double C) {
        this.R_in = R_in;
        this.C = C;
    }

    public void decay(double amount) {
        this.update(0, amount, 1);
        lastUpdatedAt = Utils.now();
    }

    public void charge(double amount) {
        this.update(1, amount, this.R_in);
    }

    public void discharge(double amount) {
        this.update(-1, amount, this.R_in);
    }


    protected void update(int target, double amount, double resistance) {
        this.value += (target - this.value) * (1 - Math.exp(-amount / (resistance * this.C)));
    }
}
